pyanaconda/modules/payloads/payload/dnf/tree_info.py

```diff
@@ -0,0 +1,429 @@
+
+            if not_found == set(self.TREE_INFO_NAMES):
+                raise NoTreeInfoError("No treeinfo metadata found (404).")
+
+        raise NoTreeInfoError("Couldn't download treeinfo metadata.")
+
+    def verify_image_base_repo(self):
+        """Verify the base repository of an ISO image.
+
+        :return: True or False
+        """
+        repo_md = self._get_base_repository() or self._get_root_repository()
+
+        if not repo_md:
+            log.debug("There is no usable repository available")
+            return False
+
+        if not repo_md.valid:
+            log.debug("There is no valid repository available.")
+            return False
+
+        return True
+
+    def get_base_repo_url(self):
+        """Return an URL of the base repository.
+
+        :return: an URL of the base repo
+        """
+        repo_md = self._get_base_repository()
+
+        if repo_md:
+            log.debug("The treeinfo defines a base repository at: %s", repo_md.path)
+            return repo_md.path
+
+        log.debug("No base repository found in the treeinfo. Using installation tree root.")
+        return self._root_path
+
+    def _get_base_repository(self):
+        """Return metadata of the base repository.
+
+        :return: an instance of TreeInfoRepoMetadata or None
+        """
+        repos = DEFAULT_REPOS
+
+        for repo_md in self.repositories:
+            if repo_md.name in repos:
+                return repo_md
+
+        return None
+
+    def _get_root_repository(self):
+        """Return metadata of the root repository.
+
+        :return: an instance of TreeInfoRepoMetadata or None
+        """
+        for repo_md in self.repositories:
+            if repo_md.relative_path == ".":
+                return repo_md
+
+        return None
+
+
+class TreeInfoRepoMetadata(object):
+    """Metadata repo object contains metadata about repository."""
+
+    def __init__(self, repo_name, tree_info, root_path):
+        """Do not instantiate this class directly.
+
+        :param repo_name: a name of the repository
+        :param tree_info: a metadata of the repository
+        :param root_path: a root path of the installation source
+        """
+        self._name = repo_name
+        self._type = tree_info.type
+        self._root_path = root_path
+        self._relative_path = tree_info.paths.repository
+
+    @property
+    def type(self):
+        """Type of the repository."""
+        return self._type
+
+    @property
+    def name(self):
+        """Name of the repository."""
+        return self._name
+
+    @property
+    def enabled(self):
+        """Is the repository enabled?
+
+        :return: True or False
+        """
+        return self._type in conf.payload.enabled_repositories_from_treeinfo
+
+    @property
+    def valid(self):
+        """Is the repository valid?
+
+        This is only check if the repository is invalid.
+        If success it doesn't mean it is valid.
```

Comment:
Funny :D, maybe we could have 'invalid' attribute then, but even that might not help. Feel free to ignore this comment.

Reply:
Fixed.